Mssql connections api

elpaso opened the pull request:
Implementation of the connections API for MSSQL

elpaso committed:
MSSQL connections API initial implementation

elpaso committed:
Return an empty geomety types list when aspatial

elpaso committed:
Add MSSQL exceptions and UTF8 tests

elpaso committed:
MSSQL connections API

nyalldawson commented on src/providers/mssql/qgsmssqlproviderconnection.cpp:
```diff
@@ -0,0 +1,482 @@
+
+    int geomColCount { 0 };
+
+    if ( ! table.geometryColumn().isEmpty() )
```
How do you feel about adding a flag "GeometryColumnDetailsOptional" ="in cases where resolving the full geometry details of a table are expensive, setting this flag will skip fetching these details"

elpaso replied:
That's a good idea, there are cases where you just need the table names and you don't care about other details. But I would leave this to a separate PR because it impacts all other providers.

nyalldawson commented on tests/src/python/test_qgsproviderconnection_base.py:
```diff
@@ -184,23 +213,30 @@ def _test_operations(self, md, conn):
-                sql = "DELETE FROM %s WHERE string = 'QGIS Rocks - \U0001f604'" % table
+
+                # This does not work in MSSQL and returns a QByteArray, we have no way to know that it is a time
+                # value and there is no way we can convert it.
```
The provider iterator has code which does this -- but I guess the issue here is that you don't have the field type information to use to detect this situation?

elpaso replied:
I've investigated this issue for hours, I think there is a bug upstream in QT or in the ODBC driver, because according to QT docs `time` must be supported. And yes, we don't have a field here we can guess the type from, we could check the QByteArray size and try a reinterpret_cast conversion but it's far from being a solid approach.

nyalldawson replied:
Yep, leave it as is I say, it's the right decision

nyalldawson reviewed (commented):
It's beautiful! 😭😀

nyalldawson commented:
Related question following on from our discussion yesterday: what do you think about adding a method to QgsProviderRegistry to retrieve a connection given a string like "postgres://my_conn_name" or "mssql://conn_name"?

elpaso commented:
> Related question following on from our discussion yesterday: what do you think about adding a method to QgsProviderRegistry to retrieve a connection given a string like "postgres://my_conn_name" or "mssql://conn_name"?

I think it's a good idea, just let's make it **always** lowercase, and get finally rid of **WFS**

We should also start a QEP about  `QgsDataSourceUri` refactoring.

elpaso committed:
MSSQL revert geometryColumnTypes and homogenize

Other providers return an "invalid" col type from geometryColumnTypes
let's keep it that way (for now)
